torch/csrc/utils/future.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+
+namespace utils {
+
+// FutureError inherits from std::exception, it can return const char* or
+// std::string error message
+class TORCH_API FutureError final : public std::exception {
+public:
+  FutureError(std::string errorMsg) : errorMsg_(std::move(errorMsg)) {}
+
+  FutureError() = default;
+
+  const char* what() const noexcept override {
+    return errorMsg_.c_str();
+  }
+
+private:
+  std::string errorMsg_;
+};
+
+// This class holds a value of type T that will be ready in the future.
+// Most implementation is copied from FutureMessage and
+// c10::ivalue::Future
+template <typename T>
+class TORCH_API Future final {
+ public:
+  using Callback =
+      std::function<void(const T&, const c10::optional<FutureError>&)>;
+
+  Future() = default;
+
+  Future(T value)
+  : completed_(true), value_(std::move(value)) {}
+
+  const T& wait() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    finished_cv_.wait(lock, [this] { return completed_.load(); });
+
+    if (error_) {
+      throw *error_;
+    }
+    return value_;
+  }
+
+  const T& waitNoThrow() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    finished_cv_.wait(lock, [this] { return completed_.load(); });
+    return value_;
+  }
+
+  T&& moveValue() && {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return std::move(value_);
+  }
+
+  void markCompleted(T value) {
+    {
```

Comment:
nit: extra scope isn't really necessary here or in setError(), since the code drops the lock early now